File: daemons/macos/src/Runner.swift
import Foundation
import Network

final class Runner {
    let sessionId: String
    private(set) var hasExited = false
    private let hasStartedBefore: Bool
    private let model: String?
    private let effort: String?
    private let queue: DispatchQueue
    var onFinish: (() -> Void)?
    private var process: Process?
    private var stdinPipe: Pipe?
    private var ring: [(seq: Int, data: Data)] = []
    private var subscribers: [NWConnection] = []
    private var seq = 0
    private let maxRingSize = 1000
    private var lineBuffer = Data()

    init(sessionId: String, hasStartedBefore: Bool, model: String?, effort: String?, queue: DispatchQueue) {
        self.sessionId = sessionId
        self.hasStartedBefore = hasStartedBefore
        self.model = model
        self.effort = effort
        self.queue = queue
    }

    func spawn(path: String, prompt: String) {
        let proc = Process()
        let stdout = Pipe()
        let stdin = Pipe()
        let stderr = Pipe()

        let executable = Runner.claudeExecutable()
        proc.executableURL = URL(fileURLWithPath: executable.path)
        var claudeArgs =
            executable.leadingArguments + [
                "-p",
                "--output-format", "stream-json",
                "--verbose",
                "--include-partial-messages",
                "--disallowedTools", "AskUserQuestion ExitPlanMode EnterPlanMode",
            ]
        if let model {
            claudeArgs.append(contentsOf: ["--model", model])
        }
        if let effort {
            claudeArgs.append(contentsOf: ["--effort", effort])
        }
        let normalizedId = sessionId.lowercased()
        if hasStartedBefore {
            claudeArgs.append(contentsOf: ["--resume", normalizedId])
        } else {
            claudeArgs.append(contentsOf: ["--session-id", normalizedId])
        }
        proc.arguments = claudeArgs
        proc.currentDirectoryURL = URL(fileURLWithPath: path)
        proc.standardInput = stdin
        proc.standardOutput = stdout
        proc.standardError = stderr
        proc.environment = Runner.spawnEnvironment()

        #if DEBUG
        NSLog(
            "[Runner] spawn sessionId=\(sessionId) hasStartedBefore=\(hasStartedBefore) promptChars=\(prompt.count) executable=\(executable.path)"
        )
        #endif

        stdout.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            if data.isEmpty {
                handle.readabilityHandler = nil
                return
            }
            self?.queue.async { self?.ingest(data) }
        }

        stderr.fileHandleForReading.readabilityHandler = { [weak self] handle in
            let data = handle.availableData
            if data.isEmpty {
                handle.readabilityHandler = nil
                return
            }
            #if DEBUG
            let text = String(data: data, encoding: .utf8) ?? "<binary \(data.count)>"
            NSLog("[Runner] stderr sessionId=\(self?.sessionId ?? "?"): \(text)")
            #endif
        }

        proc.terminationHandler = { [weak self] p in
            self?.queue.async { self?.finish(exitCode: p.terminationStatus) }
        }

        do {
            try proc.run()
            self.process = proc
            self.stdinPipe = stdin
            stdin.fileHandleForWriting.write(Data(prompt.utf8))
            try? stdin.fileHandleForWriting.close()
        } catch {
            #if DEBUG
            NSLog("[Runner] spawn_failed sessionId=\(sessionId) error=\(error)")
            #endif
            emit(["type": "error", "message": "spawn_failed: \(error.localizedDescription)"])
            finish(exitCode: -1)
        }
    }

    func subscribe(_ connection: NWConnection, afterSeq: Int = -1) {
        #if DEBUG
        NSLog(
            "[Runner] subscribe sessionId=\(sessionId) afterSeq=\(afterSeq) ringSize=\(ring.count) hasExited=\(hasExited)"
        )
        #endif
        connection.stateUpdateHandler = { [weak self, weak connection] state in
            switch state {
            case .failed, .cancelled:
                self?.queue.async { self?.removeSubscriber(connection) }
            default: break
            }
        }
        var batch = Data()
        for (s, data) in ring where s > afterSeq { batch.append(data) }
        if !batch.isEmpty {
            connection.send(content: batch, completion: .contentProcessed { _ in })
        }
        if hasExited {
            connection.cancel()
        } else {
            subscribers.append(connection)
        }
    }

    func abort() {
        if let proc = process, proc.isRunning {
            emit(["type": "aborted"])
            proc.interrupt()
        }
    }

    private func removeSubscriber(_ connection: NWConnection?) {
        if let connection = connection {
            subscribers.removeAll { $0 === connection }
        }
    }

    private func ingest(_ data: Data) {
        lineBuffer.append(data)
        while let nl = lineBuffer.firstIndex(of: 0x0A) {
            let line = lineBuffer.subdata(in: 0..<nl)
            lineBuffer.removeSubrange(0...nl)
            if line.isEmpty { continue }
            if let obj = try? JSONSerialization.jsonObject(with: line) {
                emit(["event": obj])
            } else {
                #if DEBUG
                let text = String(data: line, encoding: .utf8) ?? "<binary \(line.count)>"
                NSLog("[Runner] stdout non-json sessionId=\(sessionId): \(text)")
                #endif
            }
        }
    }

    private func emit(_ partial: [String: Any]) {
        seq += 1
        var wrapped = partial
        wrapped["seq"] = seq
        wrapped["sessionId"] = sessionId
        if let payload = try? JSONSerialization.data(withJSONObject: wrapped) {
            var chunk = payload
            chunk.append(0x0A)
            ring.append((seq, chunk))
            if ring.count > maxRingSize { ring.removeFirst(ring.count - maxRingSize) }
            for sub in subscribers {
                sub.send(content: chunk, completion: .contentProcessed { _ in })
            }
        }
    }

    private func finish(exitCode: Int32) {
        if hasExited { return }
        hasExited = true
        if let stderrPipe = process?.standardError as? Pipe {
            stderrPipe.fileHandleForReading.readabilityHandler = nil
            let remaining = (try? stderrPipe.fileHandleForReading.readToEnd()) ?? Data()
            #if DEBUG
            let text = String(data: remaining, encoding: .utf8) ?? "<binary \(remaining.count)>"
            NSLog("[Runner] stderr (drained) sessionId=\(sessionId) bytes=\(remaining.count): \(text)")
            #endif
        }
        if let stdoutPipe = process?.standardOutput as? Pipe {
            stdoutPipe.fileHandleForReading.readabilityHandler = nil
            let remaining = (try? stdoutPipe.fileHandleForReading.readToEnd()) ?? Data()
            #if DEBUG
            let text = String(data: remaining, encoding: .utf8) ?? "<binary \(remaining.count)>"
            NSLog("[Runner] stdout (drained) sessionId=\(sessionId) bytes=\(remaining.count): \(text)")
            #endif
        }
        #if DEBUG
        NSLog(
            "[Runner] finish sessionId=\(sessionId) exitCode=\(exitCode) ringSize=\(ring.count) subscribers=\(subscribers.count)"
        )
        #endif
        emit(["type": "exit", "code": Int(exitCode)])
        for sub in subscribers { sub.cancel() }
        subscribers.removeAll()
        process?.standardOutput.flatMap { ($0 as? Pipe) }?.fileHandleForReading.readabilityHandler = nil
        onFinish?()
    }

    private static func spawnEnvironment() -> [String: String] {
        let inherited = ProcessInfo.processInfo.environment
        var env: [String: String] = [:]
        for key in ["HOME", "USER", "SHELL", "LANG", "LC_ALL", "TMPDIR", "TERM"] {
            if let value = inherited[key] { env[key] = value }
        }
        let home = FileManager.default.homeDirectoryForCurrentUser.path
        var pathParts = (inherited["PATH"] ?? "").split(separator: ":").map(String.init)
        for extra in [
            "/opt/homebrew/bin",
            "/usr/local/bin",
            "/usr/bin",
            "\(home)/.local/bin",
            "\(home)/.npm-global/bin",
        ] where !pathParts.contains(extra) {
            pathParts.append(extra)
        }
        env["PATH"] = pathParts.joined(separator: ":")
        env["TERM"] = env["TERM"] ?? "xterm-256color"
        env["NO_COLOR"] = "1"
        env["CLAUDE_CODE_EXPERIMENTAL_AGENT_TEAMS"] = "1"
        return env
    }

    private struct Executable {
        let path: String
        let leadingArguments: [String]
    }

    private static func claudeExecutable() -> Executable {
        let fileManager = FileManager.default
        let home = fileManager.homeDirectoryForCurrentUser.path
        var directories = (ProcessInfo.processInfo.environment["PATH"] ?? "").split(separator: ":").map(
            String.init)
        for extra in [
            "/opt/homebrew/bin",
            "/usr/local/bin",
            "/usr/bin",
            "\(home)/.local/bin",
            "\(home)/.npm-global/bin",
        ] where !directories.contains(extra) {
            directories.append(extra)
        }
        for directory in directories {
            let candidate = "\(directory)/claude"
            if fileManager.isExecutableFile(atPath: candidate) {
                return Executable(path: candidate, leadingArguments: [])
            }
        }
        return Executable(path: "/usr/bin/env", leadingArguments: ["claude"])
    }
}

File: daemons/linux/src/Runner.js
import { spawn } from 'node:child_process'
import { StringDecoder } from 'node:string_decoder'
import { claudeCommand, spawnEnvironment } from './Runtime/ClaudeRuntime.js'

export default class Runner {
  constructor({ sessionId, hasStartedBefore, model, effort, permissionMode, onFinish }) {
    this.sessionId = sessionId
    this.hasExited = false
    this.hasStartedBefore = hasStartedBefore
    this.model = model
    this.effort = effort
    this.permissionMode = permissionMode
    this.onFinish = onFinish
    this.process = null
    this.ring = []
    this.subscribers = new Set()
    this.seq = 0
    this.lineBuffer = ''
    this.decoder = new StringDecoder('utf8')
  }

  spawn(path, prompt) {
    const { executable, leadingArguments } = claudeCommand()
    const argumentsList = [
      ...leadingArguments,
      '-p',
      '--output-format',
      'stream-json',
      '--verbose',
      '--include-partial-messages',
      '--disallowedTools',
      'AskUserQuestion ExitPlanMode EnterPlanMode Monitor ScheduleWakeup CronCreate CronDelete CronList RemoteTrigger PushNotification TeamCreate TeamDelete'
    ]
    const permissionFlags = new Map([
      ['plan', ['--permission-mode', 'plan']],
      ['default', ['--permission-mode', 'default']],
      ['acceptEdits', ['--permission-mode', 'acceptEdits']],
      ['custom', []]
    ])
    argumentsList.push(...(permissionFlags.get(this.permissionMode) ?? ['--dangerously-skip-permissions']))
    if (this.model) {
      argumentsList.push('--model', this.model)
    }
    if (this.effort) {
      argumentsList.push('--effort', this.effort)
    }
    if (this.hasStartedBefore) {
      argumentsList.push('--resume', this.sessionId.toLowerCase())
    } else {
      argumentsList.push('--session-id', this.sessionId.toLowerCase())
    }
    this.process = spawn(executable, argumentsList, {
      cwd: path,
      env: spawnEnvironment(),
      stdio: ['pipe', 'pipe', 'pipe']
    })
    this.process.stdout.on('data', (data) => {
      this.ingest(this.decoder.write(data))
    })
    this.process.stderr.on('data', (data) => {
      console.error(`Runner[${this.sessionId}]: ${data}`)
    })
    this.process.on('error', (error) => {
      this.emit({ type: 'error', message: `spawn_failed: ${error.message}` })
      this.finish(-1)
    })
    this.process.on('close', (code) => {
      this.finish(typeof code === 'number' ? code : -1)
    })
    if (this.process.stdin) {
      this.process.stdin.on('error', (error) => {
        console.error(`Runner[${this.sessionId}] stdin: ${error.message}`)
      })
      this.process.stdin.write(prompt)
      this.process.stdin.end()
    }
  }

  subscribe(response, afterSeq = -1) {
    response.on('close', () => {
      this.subscribers.delete(response)
    })
    response.on('error', () => {
      this.subscribers.delete(response)
    })
    for (const entry of this.ring) {
      if (entry.seq > afterSeq) {
        response.write(entry.data)
      }
    }
    if (this.hasExited) {
      response.end()
      return
    }
    this.subscribers.add(response)
  }

  abort() {
    if (this.process && this.process.exitCode === null) {
      this.emit({ type: 'aborted' })
      this.process.kill('SIGINT')
      setTimeout(() => {
        if (!this.hasExited) {
          this.process.kill('SIGKILL')
        }
      }, 5000).unref()
    }
  }

  ingest(chunk) {
    this.lineBuffer += chunk
    let newline = this.lineBuffer.indexOf('\n')
    while (newline !== -1) {
      const line = this.lineBuffer.slice(0, newline)
      this.lineBuffer = this.lineBuffer.slice(newline + 1)
      if (line.length !== 0) {
        let parsed = null
        try {
          parsed = JSON.parse(line)
        } catch {
          parsed = null
        }
        if (parsed) {
          if (parsed.type === 'system' && parsed.subtype === 'informational' && parsed.status === 'compacting') {
            this.emit({ type: 'status', state: 'compacting' })
          } else {
            this.emit({ event: parsed })
          }
        }
      }
      newline = this.lineBuffer.indexOf('\n')
    }
  }

  emit(partial) {
    this.seq += 1
    const data = Buffer.from(
      `${JSON.stringify({ ...partial, seq: this.seq, sessionId: this.sessionId })}\n`
    )
    this.ring.push({ seq: this.seq, data })
    if (this.ring.length > 1000) {
      this.ring.splice(0, this.ring.length - 1000)
    }
    for (const subscriber of this.subscribers) {
      subscriber.write(data)
      if (subscriber.writableLength > 8 * 1024 * 1024) {
        subscriber.destroy()
      }
    }
  }

  finish(exitCode) {
    if (this.hasExited) {
      return
    }
    this.lineBuffer += this.decoder.end()
    if (this.lineBuffer.trim().length !== 0) {
      let parsed = null
      try {
        parsed = JSON.parse(this.lineBuffer)
      } catch {
        parsed = null
      }
      if (parsed) {
        this.emit({ event: parsed })
      }
    }
    this.hasExited = true
    this.emit({ type: 'exit', code: exitCode })
    for (const subscriber of this.subscribers) {
      subscriber.end()
    }
    this.subscribers.clear()
    this.onFinish?.()
  }
}
